File: javascript/src/rag/ingest.ts
import { Chroma } from "@langchain/community/vectorstores/chroma";
import { OpenAIEmbeddings } from "@langchain/openai";
import { CheerioWebBaseLoader } from "@langchain/community/document_loaders/web/cheerio";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import dotenv from "dotenv";
import { Document } from "langchain/document";
dotenv.config();

async function ingestDocuments() {
  // 3. Initialize embeddings model
  const embeddings = new OpenAIEmbeddings();

  // 1. Load documents
  const urls = [
    "https://lilianweng.github.io/posts/2023-06-23-agent/",
    "https://lilianweng.github.io/posts/2023-03-15-prompt-engineering/",
  ];
  const docSplits: Document<Record<string, any>>[] = [];
  for (const url of urls) {
    const loader = new CheerioWebBaseLoader(url);
    const docs = await loader.load();

    // 2. Split text into chunks
    const textSplitter = new RecursiveCharacterTextSplitter({
      chunkSize: 500,
      chunkOverlap: 100,
    });
    const ds = await textSplitter.splitDocuments(docs);
    docSplits.push(...ds);
  }
  // 4. Create and persist Chroma vector store
  await Chroma.fromDocuments(docSplits, embeddings, {
    collectionName: "ai_blog_posts",
    url: "http://localhost:8000", // Chroma server URL
  });
}

async function main() {
  await ingestDocuments();
}

main();

File: python/langchain/rag/ingest.py
from langchain_chroma import Chroma
from langchain_openai import OpenAIEmbeddings
from langchain_community.document_loaders import WebBaseLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_docling import DoclingLoader

# 1. ドキュメントの読み込み
urls = [
    "https://lilianweng.github.io/posts/2023-06-23-agent/",
    "https://lilianweng.github.io/posts/2023-03-15-prompt-engineering/",
    "https://www.anthropic.com/research/building-effective-agents",
]
loader = WebBaseLoader(urls)
docs = loader.load()

# テキストの分割
text_splitter = RecursiveCharacterTextSplitter(chunk_size=5000, chunk_overlap=1000)
doc_splits = text_splitter.split_documents(docs)

# 埋め込みモデルの初期化
embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

# Chromaベクターストアの作成
# FAISS
# PINECONE
vectorstore = Chroma(
    collection_name="ai_blog_posts",
    embedding_function=embeddings,
    persist_directory="chroma_db",
)

# ドキュメントの保存
vectorstore.add_documents(doc_splits)
